File: ios/NMM/Data/FinanceRepository.swift
import Foundation
import SwiftData

@MainActor
final class FinanceRepository {
    private let modelContext: ModelContext

    init(modelContext: ModelContext) {
        self.modelContext = modelContext
    }

    // MARK: - Transactions

    func fetchAllTransactions() throws -> [Transaction] {
        let descriptor = FetchDescriptor<Transaction>(sortBy: [SortDescriptor(\.date, order: .reverse)])
        return try modelContext.fetch(descriptor)
    }

    func insertTransaction(_ transaction: Transaction) throws {
        modelContext.insert(transaction)
        try modelContext.save()
    }

    func deleteTransaction(_ transaction: Transaction) throws {
        modelContext.delete(transaction)
        try modelContext.save()
    }

    // MARK: - Budgets

    func fetchAllBudgets() throws -> [Budget] {
        let descriptor = FetchDescriptor<Budget>(sortBy: [SortDescriptor(\.category)])
        return try modelContext.fetch(descriptor)
    }

    func insertBudget(_ budget: Budget) throws {
        modelContext.insert(budget)
        try modelContext.save()
    }

    func deleteBudget(_ budget: Budget) throws {
        modelContext.delete(budget)
        try modelContext.save()
    }

    func budget(for category: String) throws -> Budget? {
        let descriptor = FetchDescriptor<Budget>(
            predicate: #Predicate { $0.category == category }
        )
        return try modelContext.fetch(descriptor).first
    }

    // MARK: - Savings Goals

    func fetchAllSavingsGoals() throws -> [SavingsGoal] {
        let descriptor = FetchDescriptor<SavingsGoal>(sortBy: [SortDescriptor(\.deadline)])
        return try modelContext.fetch(descriptor)
    }

    func insertSavingsGoal(_ goal: SavingsGoal) throws {
        modelContext.insert(goal)
        try modelContext.save()
    }

    func deleteSavingsGoal(_ goal: SavingsGoal) throws {
        modelContext.delete(goal)
        try modelContext.save()
    }

    // MARK: - Reset

    func resetDatabase() throws {
        try fetchAllTransactions().forEach { modelContext.delete($0) }
        try fetchAllBudgets().forEach { modelContext.delete($0) }
        try fetchAllSavingsGoals().forEach { modelContext.delete($0) }
        try modelContext.save()
    }
}

File: app/src/main/java/com/example/data/repository/FinanceRepository.kt
package com.example.data.repository

import com.example.data.db.FinanceDao
import com.example.data.model.Budget
import com.example.data.model.SavingsGoal
import com.example.data.model.Transaction
import kotlinx.coroutines.flow.Flow

class FinanceRepository(private val financeDao: FinanceDao) {

    // Transactions
    val allTransactions: Flow<List<Transaction>> = financeDao.getAllTransactions()

    suspend fun getTransactionById(id: Long): Transaction? {
        return financeDao.getTransactionById(id)
    }

    suspend fun insertTransaction(transaction: Transaction) {
        financeDao.insertTransaction(transaction)
    }

    suspend fun deleteTransaction(transaction: Transaction) {
        financeDao.deleteTransaction(transaction)
    }

    suspend fun resetDatabase() {
        financeDao.clearAllTransactions()
        financeDao.clearAllBudgets()
        financeDao.clearAllSavingsGoals()
    }

    // Budgets
    val allBudgets: Flow<List<Budget>> = financeDao.getAllBudgets()

    suspend fun getBudgetByCategory(category: String): Budget? {
        return financeDao.getBudgetByCategory(category)
    }

    suspend fun insertBudget(budget: Budget) {
        financeDao.insertBudget(budget)
    }

    suspend fun deleteBudget(budget: Budget) {
        financeDao.deleteBudget(budget)
    }

    // Savings Goals
    val allSavingsGoals: Flow<List<SavingsGoal>> = financeDao.getAllSavingsGoals()

    suspend fun getSavingsGoalById(id: Long): SavingsGoal? {
        return financeDao.getSavingsGoalById(id)
    }

    suspend fun insertSavingsGoal(savingsGoal: SavingsGoal) {
        financeDao.insertSavingsGoal(savingsGoal)
    }

    suspend fun deleteSavingsGoal(savingsGoal: SavingsGoal) {
        financeDao.deleteSavingsGoal(savingsGoal)
    }
}
